src/scene/particle-emitter.js

```diff
@@ -1526,31 +1523,86 @@ Object.assign(pc, function () {
             // #endif
         },
 
-        destroy: function () {
-            if (this.particleTexIN) this.particleTexIN.destroy();
-            if (this.particleTexOUT) this.particleTexOUT.destroy();
-            if (!this.useCpu && this.particleTexStart) this.particleTexStart.destroy();
-            if (this.rtParticleTexIN) this.rtParticleTexIN.destroy();
-            if (this.rtParticleTexOUT) this.rtParticleTexOUT.destroy();
-
-            // TODO: delete shaders from cache with reference counting
-            // if (this.shaderParticleUpdateRespawn) this.shaderParticleUpdateRespawn.destroy();
-            // if (this.shaderParticleUpdateNoRespawn) this.shaderParticleUpdateNoRespawn.destroy();
-            // if (this.shaderParticleUpdateOnStop) this.shaderParticleUpdateOnStop.destroy();
-
-            this.particleTexIN = null;
-            this.particleTexOUT = null;
-            this.particleTexStart = null;
-            this.rtParticleTexIN = null;
-            this.rtParticleTexOUT = null;
-
-            this.shaderParticleUpdateRespawn = null;
-            this.shaderParticleUpdateNoRespawn = null;
-            this.shaderParticleUpdateOnStop = null;
+        _destroyResources: function () {
+            if(this.particleTexIN) {
+                this.particleTexIN.destroy();
+                this.particleTexIN = null;
+            }
+
+            if (this.particleTexOUT) {
+                this.particleTexOUT.destroy();
+                this.particleTexOUT = null;
+            }
+
+            if (this.particleTexStart && this.particleTexStart.destroy) {
+                this.particleTexStart.destroy();
+                this.particleTexStart = null;
+            }
+
+            if (this.rtParticleTexIN) {
+                this.rtParticleTexIN.destroy();
+                this.rtParticleTexIN = null;
+            }
+
+            if (this.rtParticleTexOUT) {
+                this.rtParticleTexOUT.destroy();
+                this.rtParticleTexOUT = null;
+            }
+
+            if (this.internalTex0) {
+                this.internalTex0.destroy();
+                this.internalTex0 = null;
+            }
+
+            if (this.internalTex1) {
+                this.internalTex1.destroy();
+                this.internalTex1 = null;
+            }
+
+            if (this.internalTex2) {
+                this.internalTex2.destroy();
+                this.internalTex2 = null;
+            }
+
+            if (this.internalTex3) {
+                this.internalTex3.destroy();
+                this.internalTex3 = null;
+            }
+
+            if (this.shaderParticleUpdateRespawn) {
+                this.shaderParticleUpdateRespawn.destroy();
```

Comment:
@willeastcott looks like you had chosen not to do this in the old code, remember why?